internal/util/flowgraph/node.go

```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
+		default:
+			// inputs from inputsMessages for Operate
+			var input, output []Msg
+			// inputNode.input not from nodeCtx.inputChannel
+			// the input message decides whether the operate method is executed
+			n := inputNode.node
+			inputNode.blockMutex.RLock()
+			if !n.IsValidInMsg(input) {
+				inputNode.blockMutex.RUnlock()
+				return
+			}
+			output = n.Operate(input)
+			inputNode.blockMutex.RUnlock()
+			// the output decide whether the node should be closed.
+			if isCloseMsg(output) {
+				close(nodeCtxManager.inputNodeCloseCh)
+				if inputNode.inputChannel != nil {
+					close(inputNode.inputChannel)
+				}
+				inputNode.Close()
+			}
+			// deliver to all following flow graph node.
+			inputNode.downstream.inputChannel <- output
+			if enableTtChecker {
+				checker.Check(fmt.Sprintf("nodeCtxTtChecker-%s", inputNode.node.Name()))
+			}
+		}
+	}
+}
+
+func (nodeCtxManager *nodeCtxManager) ddNodeWork(checker *timerecord.GroupChecker) {
+	defer nodeCtxManager.closeWg.Done()
+	ddNode := nodeCtxManager.inputNodeCtx.downstream
+	for {
+		curNode := ddNode
+		select {
+		case <-nodeCtxManager.ddNodeCloseCh:
+			// remove curNode checker
+			for curNode != nil {
+				checker.Remove(fmt.Sprintf("nodeCtxTtChecker-%s", curNode.node.Name()))
+				curNode = curNode.downstream
+			}
+			return
+		// handles node work spinning
+		// 1. collectMessage from upstream or just produce Msg from InputNode
+		// 2. invoke node.Operate
+		// 3. deliver the Operate result to downstream nodes
+		default:
+			// goroutine will work loop for all node(expect inpuNode) even when closeCh notify to exit
+			// input node will close all node
+			for curNode != nil {
+				// inputs from inputsMessages for Operate
+				var input, output []Msg
+				input = <-curNode.inputChannel
+				// the input message decides whether the operate method is executed
+				n := curNode.node
+				curNode.blockMutex.RLock()
+				if !n.IsValidInMsg(input) {
+					curNode.blockMutex.RUnlock()
+					return
+				}
+				output = n.Operate(input)
+				curNode.blockMutex.RUnlock()
+				// the output decide whether the node should be closed.
+				if isCloseMsg(output) {
+					nodeCtxManager.closeOnce.Do(func() {
+						close(nodeCtxManager.ddNodeCloseCh)
+					})
+					if curNode.inputChannel != nil {
+						close(curNode.inputChannel)
+					}
+				}
+				// deliver to all following flow graph node.
+				if curNode.downstream != nil {
+					curNode.downstream.inputChannel <- output
+				}
+				if enableTtChecker {
+					checker.Check(fmt.Sprintf("nodeCtxTtChecker-%s", curNode.node.Name()))
+				}
+				curNode = curNode.downstream
+			}
+		}
+	}
+}
+
+// Close handles cleanup logic and notify worker to quit
+func (nodeCtxManager *nodeCtxManager) Close() {
+	nodeCtx := nodeCtxManager.inputNodeCtx
+	for nodeCtx != nil {
+		nodeCtx.node.Close()
+		log.Debug("flow graph node closed", zap.String("nodeName", nodeCtx.node.Name()))
+		nodeCtx = nodeCtx.downstream
```

Comment:
this line does nothing, it assign a new address to the pointer `nodeCtx`, but not `nodeCtxManager.intputNodeCtx`. They are only two different pointer variables which point to the same address.

Also the closing way is not thread safe, and too tricky. Implementing the closing logic in node itself, the manager should only call the `Close()` of nodes, and wait it done